File: companion-ios/Packages/OBCKit/Sources/OBCProtocolV3/ErrorBody.swift
import Foundation

/// §12's presence bits.
public struct ErrorPresence: OptionSet, Sendable, Hashable {
    public let rawValue: UInt16
    public init(rawValue: UInt16) { self.rawValue = rawValue }
    public static let retryDelay = ErrorPresence(rawValue: 1 << 0)
    public static let expectedOffset = ErrorPresence(rawValue: 1 << 1)
    public static let currentRevision = ErrorPresence(rawValue: 1 << 2)
    public static let requiredBytes = ErrorPresence(rawValue: 1 << 3)
    public static let availableBytes = ErrorPresence(rawValue: 1 << 4)
    /// §12: a durable claim exists for this OperationId under this principal.
    public static let durableClaimExists = ErrorPresence(rawValue: 1 << 5)
    /// §12: meaningful only with bit 5 — that claim is now terminal.
    public static let claimIsTerminal = ErrorPresence(rawValue: 1 << 6)
    static let defined: UInt16 = 0x007F
}

/// The 48-byte prefix plus optional non-authoritative diagnostic text of §12.
///
/// Decoding is deliberately permissive about *content*: §12's presence and guidance requirements
/// "bind senders only", so this never rejects a body because an optional field is present where it
/// expected none or absent where the category would normally require one. That permissiveness is
/// exactly what makes §11's replayed terminal bodies decodable without a special case. Only the
/// structural rules are enforced.
public struct ErrorBody: Hashable, Sendable {
    public static let prefixBytes = 48

    public let rawCategory: UInt16
    /// Detail namespace: common `0` or the affected ObjectKind for `semanticValidation`.
    public let namespace: UInt16
    public let detail: UInt16
    public let rawGuidance: UInt8
    public let rawOwner: UInt8
    public let presence: ErrorPresence
    public let retryAfterMilliseconds: UInt32
    public let expectedOffset: UInt64
    public let currentRevision: UInt64
    public let requiredBytes: UInt64
    public let availableBytes: UInt64
    /// Raw text bytes. §12: "A receiver MUST NOT reject a frame because its diagnostic text is
    /// malformed" — the bytes are kept verbatim and rendered lossily on demand.
    public let text: [UInt8]

    public var category: ErrorCategory? { ErrorCategory(rawValue: rawCategory) }
    public var guidance: RetryGuidance? { RetryGuidance(rawValue: rawGuidance) }
    public var owner: ErrorOwner? { ErrorOwner(rawValue: rawOwner) }
    public var encodedLength: Int { Self.prefixBytes + text.count }

    /// §12: rendered lossily, never parsed, matched on, or turned into behaviour.
    public var lossyText: String { String(decoding: text, as: UTF8.self) }

    /// §11: a retained Aborted replay carries both status bits and forced reject-permanently
    /// guidance. This is the discriminator a client may test to recognise one.
    public var looksLikeRetainedTerminalReplay: Bool {
        presence.contains(.durableClaimExists) && presence.contains(.claimIsTerminal)
    }

    public var fault: WireFault {
        WireFault(category ?? .internal, detail, namespace: namespace)
    }

    public init(
        rawCategory: UInt16, namespace: UInt16 = 0, detail: UInt16, rawGuidance: UInt8,
        rawOwner: UInt8 = 0, presence: ErrorPresence = [], retryAfterMilliseconds: UInt32 = 0,
        expectedOffset: UInt64 = 0, currentRevision: UInt64 = 0, requiredBytes: UInt64 = 0,
        availableBytes: UInt64 = 0, text: [UInt8] = []
    ) {
        self.rawCategory = rawCategory
        self.namespace = namespace
        self.detail = detail
        self.rawGuidance = rawGuidance
        self.rawOwner = rawOwner
        self.presence = presence
        self.retryAfterMilliseconds = retryAfterMilliseconds
        self.expectedOffset = expectedOffset
        self.currentRevision = currentRevision
        self.requiredBytes = requiredBytes
        self.availableBytes = availableBytes
        self.text = text
    }

    public static func decode(_ bytes: [UInt8]) throws -> ErrorBody {
        guard bytes.count >= prefixBytes else {
            throw WireFault.truncated("ErrorBody: \(bytes.count) bytes")
        }
        var reader = ByteReader(bytes, subject: "ErrorBody")
        let rawCategory = try reader.u16()
        let namespace = try reader.u16()
        let detail = try reader.u16()
        let rawGuidance = try reader.u8()
        let rawOwner = try reader.u8()
        let presenceRaw = try reader.u16()
        let retryAfter = try reader.u32()
        let expectedOffset = try reader.u64()
        let currentRevision = try reader.u64()
        let requiredBytes = try reader.u64()
        let availableBytes = try reader.u64()
        let textLength = Int(try reader.u8())
        try reader.reserved(1, "byte 47")

        // §12: "Category 0 is reserved and invalid. A sender never emits it and a receiver treats
        // it as a malformed body rather than as an unknown future category." A *nonzero* unknown
        // category is therefore kept, not rejected.
        guard rawCategory != 0 else { throw WireFault.unknownEnum("ErrorBody: category 0") }
        if let category = ErrorCategory(rawValue: rawCategory), category != .semanticValidation,
            namespace != 0
        {
            throw WireFault.invalidCombination(
                "ErrorBody: \(category.name) with namespace \(namespace)")
        }
        guard presenceRaw & ~ErrorPresence.defined == 0 else {
            throw WireFault.reservedBits("ErrorBody: presence bits 7…15")
        }
        let presence = ErrorPresence(rawValue: presenceRaw)
        guard !(presence.contains(.claimIsTerminal) && !presence.contains(.durableClaimExists))
        else {
            throw WireFault.invalidCombination("ErrorBody: terminal bit without the claim bit")
        }
        // §12: "Only the text length field is structural: a length above 64, or a length that
        // disagrees with the frame's payload length, is `invalidFrame` as usual."
        guard textLength <= WireLimits.errorTextCeiling else {
            throw WireFault.payloadLength("ErrorBody: text length \(textLength)")
        }
        guard reader.remaining == textLength else {
            throw WireFault.payloadLength(
                "ErrorBody: text length \(textLength), \(reader.remaining) byte(s) present")
        }
        return ErrorBody(
            rawCategory: rawCategory, namespace: namespace, detail: detail,
            rawGuidance: rawGuidance, rawOwner: rawOwner, presence: presence,
            retryAfterMilliseconds: retryAfter, expectedOffset: expectedOffset,
            currentRevision: currentRevision, requiredBytes: requiredBytes,
            availableBytes: availableBytes, text: reader.rest())
    }

    public func encoded() throws -> [UInt8] {
        try requireAtMost(text.count, WireLimits.errorTextCeiling, "ErrorBody: diagnostic text")
        var writer = ByteWriter()
        writer.u16(rawCategory)
        writer.u16(namespace)
        writer.u16(detail)
        writer.u8(rawGuidance)
        writer.u8(rawOwner)
        writer.u16(presence.rawValue)
        writer.u32(retryAfterMilliseconds)
        writer.u64(expectedOffset)
        writer.u64(currentRevision)
        writer.u64(requiredBytes)
        writer.u64(availableBytes)
        writer.u8(try narrowU8(text.count, "ErrorBody: text length"))
        writer.u8(0)
        writer.raw(text)
        return writer.bytes
    }
}

File: builder/app/src/lib/dosv3/errorBody.ts
/**
 * ErrorBody, Device_Object_Protocol_v3.md §12.
 *
 * Two rules shape this decoder more than the byte table does.
 *
 * The presence matrix **binds senders only**: "A decoder MUST NOT reject an ErrorBody because an
 * optional field is present where it expected none, or absent where the category would normally
 * require one." That is what makes the retained-Aborted replays of §11 — durable records whose
 * presence bits are forced clear and whose guidance is forced to reject-permanently — decodable
 * without a special case, so nothing here checks a category against its required presence.
 *
 * Diagnostic text is **never** a reason to reject: "refusing an error body would destroy the only
 * report of a real failure to protect a field that drives nothing." Only its length field is
 * structural. So the text is carried as bytes, rendered lossily on demand, and an invalid-UTF-8
 * body decodes.
 */

import { Cursor, Writer } from "./bytes";
import { renderDiagnosticText } from "./metadata";
import { OBJECT_KIND_NAME, semanticDetailName, type ObjectKindName } from "./registry";
import {
    CATEGORY,
    CATEGORY_NAME,
    MAX_GUIDANCE,
    MAX_OWNER,
    detailName,
    reject,
    type CategoryName,
    type DosError,
} from "./result";

export const ERROR_BODY_PREFIX_BYTES = 48;
export const MAX_ERROR_TEXT_BYTES = 64;

/** §12 presence bits. Bits 7..15 are zero. */
export const PRESENCE = {
    retryDelay: 1 << 0,
    expectedOffset: 1 << 1,
    currentRevision: 1 << 2,
    requiredBytes: 1 << 3,
    availableBytes: 1 << 4,
    durableClaimExists: 1 << 5,
    claimIsTerminal: 1 << 6,
} as const;
export const PRESENCE_MASK = 0x7f;

export interface ErrorBody {
    readonly categoryValue: number;
    /** The registered name, or "unknown" for a category above the v3.0 table. */
    readonly category: CategoryName | "unknown";
    /** Detail namespace: common `0`, or the ObjectKind that owns a semanticValidation rule. */
    readonly namespace: number;
    readonly namespaceKind?: ObjectKindName;
    readonly detailValue: number;
    readonly detail: string;
    readonly guidance: number;
    readonly owner: number;
    readonly presence: number;
    readonly retryAfterMs: number;
    readonly expectedOffset: bigint;
    readonly currentRevision: bigint;
    readonly requiredBytes: bigint;
    readonly availableBytes: bigint;
    readonly text: Uint8Array;
}

export function decodeErrorBody(bytes: Uint8Array): ErrorBody {
    const cursor = new Cursor(bytes);
    const categoryValue = cursor.u16();
    const namespace = cursor.u16();
    const detailValue = cursor.u16();
    const guidance = cursor.u8();
    const owner = cursor.u8();
    const presence = cursor.u16();
    const retryAfterMs = cursor.u32();
    const expectedOffset = cursor.u64();
    const currentRevision = cursor.u64();
    const requiredBytes = cursor.u64();
    const availableBytes = cursor.u64();
    const textLength = cursor.u8();
    cursor.zeros(1, "ErrorBody byte 47");

    if (categoryValue === 0) {
        reject("invalidDescriptor", "unknownEnum", "category 0 is reserved and invalid, not an unknown future one");
    }
    if (guidance > MAX_GUIDANCE) reject("invalidDescriptor", "unknownEnum", `retry guidance ${guidance} is not registered`);
    if (owner > MAX_OWNER) reject("invalidDescriptor", "unknownEnum", `owner ${owner} is not registered`);
    if ((presence & ~PRESENCE_MASK) !== 0) reject("invalidDescriptor", "reservedBits", "presence bits 7..15 are zero");
    if ((presence & PRESENCE.claimIsTerminal) !== 0 && (presence & PRESENCE.durableClaimExists) === 0) {
        reject("invalidDescriptor", "invalidCombination", "claim-is-terminal is meaningful only with durable-claim");
    }

    const category = CATEGORY_NAME.get(categoryValue) ?? "unknown";
    let namespaceKind: ObjectKindName | undefined;
    if (namespace !== 0) {
        if (categoryValue !== CATEGORY.semanticValidation) {
            reject("invalidDescriptor", "invalidCombination", "categories other than semanticValidation use namespace 0");
        }
        namespaceKind = OBJECT_KIND_NAME.get(namespace);
        if (namespaceKind === undefined) {
            reject("invalidDescriptor", "unknownEnum", `detail namespace ${namespace} is not a registered ObjectKind`);
        }
    }

    if (textLength > MAX_ERROR_TEXT_BYTES) {
        reject("invalidFrame", "payloadLength", `error text is at most ${MAX_ERROR_TEXT_BYTES} bytes`);
    }
    if (cursor.remaining !== textLength) {
        reject("invalidFrame", "payloadLength", "the text length disagrees with the body length");
    }
    const text = cursor.take(textLength);

    const detail =
        namespaceKind !== undefined
            ? semanticDetailName(namespaceKind, detailValue)
            : category === "unknown"
              ? "unknown"
              : detailName(category, detailValue);

    return {
        categoryValue,
        category,
        namespace,
        namespaceKind,
        detailValue,
        detail,
        guidance,
        owner,
        presence,
        retryAfterMs,
        expectedOffset,
        currentRevision,
        requiredBytes,
        availableBytes,
        text,
    };
}

export function encodeErrorBody(body: ErrorBody): Uint8Array {
    return new Writer(ERROR_BODY_PREFIX_BYTES + body.text.length)
        .u16(body.categoryValue)
        .u16(body.namespace)
        .u16(body.detailValue)
        .u8(body.guidance)
        .u8(body.owner)
        .u16(body.presence)
        .u32(body.retryAfterMs)
        .u64(body.expectedOffset)
        .u64(body.currentRevision)
        .u64(body.requiredBytes)
        .u64(body.availableBytes)
        .u8(body.text.length)
        .zeros(1)
        .raw(body.text)
        .finish();
}

/** True when the body carries the §11 signature of a replayed terminal result: bits 5 and 6 set. */
export const isRetainedTerminalReplay = (body: ErrorBody): boolean =>
    (body.presence & (PRESENCE.durableClaimExists | PRESENCE.claimIsTerminal)) ===
    (PRESENCE.durableClaimExists | PRESENCE.claimIsTerminal);

/** The lossy rendering §12 mandates. Never parsed, never matched on, never behaviour-bearing. */
export const errorText = (body: ErrorBody): string => renderDiagnosticText(body.text);

/** Projects a decoded ErrorBody onto this module's own failure vocabulary. */
export function asDosError(body: ErrorBody): DosError {
    return {
        category: body.category === "unknown" ? "internal" : body.category,
        categoryValue: body.categoryValue,
        detail: body.detail,
        detailValue: body.detailValue,
        message: errorText(body),
        namespace: body.namespace,
    };
}
